### PR TITLE
Added PV Fleets QA pipeline examples

### LINKED ISSUE #201: Power, irradiance, and temperature examples from PV Fleets pipeline?
Hey @kandersolar and @cwhanse since we have the logic for time shifts now merged into main, I wanted to ask if we'd be interested in adding some example notebooks illustrating the QA process that we've been using in Fleets (for temperature, irradiance, and power/energy streams):

https://github.com/kperrynrel/fleets_qa_examples/tree/main 

I know we want to keep this repo pipeline-agnostic, so these would just act as a couple of examples which illustrate how our group is using PVAnalytics functions in our work. I still need to integrate the new time shift logic into these examples (it's literally plugging in the new functions), but these could be just jupyter notebook examples, or I can integrate them as examples into Sphinx. No worries if we're not interested, but as I'm pointing multiple people towards this example repo, it could be valuable to include. Let me know your thoughts!

### PR DESCRIPTION
**Pull request by kperrynrel**
- [x] Closes #201 
- [x] Non-API functions clearly documented with docstrings or comments as necessary.
- [x] Adds description and name entries in the appropriate "what's new" file 
      in [`docs/whatsnew`](https://github.com/pvlib/pvanalytics/tree/main/docs/whatsnew) 
      for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` 
      or this Pull Request with `` :pull:`num` ``. Includes contributor name 
      and/or GitHub username (link with `` :ghuser:`user` ``).
- [x] Pull request is nearly complete and ready for detailed review.
- [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

This PR includes QA examples of the PV fleets pipeline, adapted from the following repo:  https://github.com/kperrynrel/fleets_qa_examples

### CODE REVIEW

**Comment by cwhanse** [comment]
@kperrynrel I think you'll need a README.rst file in order for the section of the Examples to be created. I would call this group of examples PVFleets QA pipeline.

**Comment by kandersolar** [comment]
The test failures can be ignored here; they'll be fixed in #203.

**Comment by kperrynrel** [comment]
@kandersolar these are ready for review. Do we have a limit on example data size? The files are bigger for these examples and I could host them on the Duramat datahub if need be. Let me know!

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+                                          decimals=2)
+stale_data_mask.loc[(stale_data_mask) &
+                    (~daytime_mask)] = False
+
+# 2) REMOVE NEGATIVE DATA
+negative_mask = (time_series < 0)
+
+# FIND ABNORMAL PERIODS
+daily_min = time_series.resample('D').min()
+erroneous_mask = (daily_min < 50)
```
Comment (cwhanse) [inline docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py:74]: This is unfortunate, would be more consistent if `True` meant good data. Maybe `daily_min > 50`

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+                      modeled_midday_series.resample('D').mean()
+                      ).dt.total_seconds() / 60
+
+# Generate boolean for detected time shifts
+if any(time_shift_series != 0):
+    time_shifts_detected = True
+else:
+    time_shifts_detected = False
+
+# Build a list of dictionaries for time shifts
```
Comment (cwhanse) [inline docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py:238]:
```suggestion
# Build a list of time shifts for re-indexing. We choose to use dicts.
```

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+# detect data shifts in the time series.
+
+# Resample the time series to daily mean
+time_series_daily = time_series.resample('D').mean()
+data_shift_start_date, data_shift_end_date = \
+    ds.get_longest_shift_segment_dates(time_series_daily)
+data_shift_period_length = (data_shift_end_date - data_shift_start_date).days
+
+# Get the number of shift dates
+data_shift_mask = pvanalytics.quality.data_shifts.detect_data_shifts(
```
Comment (cwhanse) [inline docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py:320]: `pvanalytics.quality.data_shifts` was imported as `ds` earlier.

File: docs/examples/pvfleets-qa-pipeline/pvfleets-temperature-qa.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+# Resample the time series to daily mean
+time_series_daily = time_series.resample('D').mean()
+data_shift_start_date, data_shift_end_date = \
+    ds.get_longest_shift_segment_dates(time_series_daily)
+data_shift_period_length = (data_shift_end_date -
+                            data_shift_start_date).days
+
+# Get the number of shift dates
+data_shift_mask = pvanalytics.quality.data_shifts.detect_data_shifts(
```
Comment (cwhanse) [inline docs/examples/pvfleets-qa-pipeline/pvfleets-temperature-qa.py:194]: `ds.detect_data_shifts`

File: docs/examples/pvfleets-qa-pipeline/pvfleets-temperature-qa.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+plt.tight_layout()
+plt.show()
+
+# %%
+# Now, let's filter out any of the flagged data from the basic temperature
+# checks (stale or abnormal data). Then we can re-visualize the data
+# post-filtering.
+
+# Filter the time series, taking out all of the issues
+time_series = time_series[~stale_data_mask]
```
Comment (cwhanse) [inline docs/examples/pvfleets-qa-pipeline/pvfleets-temperature-qa.py:141]: Same comment as in the irradiance example: do these filters change the length of the `Series`?

File: docs/examples/pvfleets-qa-pipeline/pvfleets-power-qa.py [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+from pvanalytics.system import (is_tracking_envelope,
+                                infer_orientation_fit_pvwatts)
+from pvanalytics.features.daytime import power_or_irradiance
+from pvanalytics.quality.time import shifts_ruptures
+from pvanalytics.features import daytime
+import pvlib
+from pvanalytics.features.clipping import geometric
+
+# %%
+# First, we import an Ac power data stream from a PV installation
```
Comment (cwhanse) [inline docs/examples/pvfleets-qa-pipeline/pvfleets-power-qa.py:31]:
```suggestion
# First, we import an AC power data stream from a PV installation
```

File: docs/examples/pvfleets-qa-pipeline/pvfleets-temperature-qa.py (hunk 4 above)
Comment (kperrynrel) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-temperature-qa.py]: Addressed in all scripts in https://github.com/pvlib/pvanalytics/pull/202/commits/cea4fbb93f97e7044f6b50102e1c8bdcf90cd7f4

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py (hunk 3 above)
Comment (kperrynrel) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py]: addressed in https://github.com/pvlib/pvanalytics/pull/202/commits/cea4fbb93f97e7044f6b50102e1c8bdcf90cd7f4

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py (hunk 1 above)
Comment (kperrynrel) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py]: updated in https://github.com/pvlib/pvanalytics/pull/202/commits/d3c2d56f7b7ced1c06cc9da4af494cb38fddc84c

**Comment by kandersolar** [comment]
> Do we have a limit on example data size? The files are bigger for these examples and I could host them on the Duramat datahub if need be.

Looks like adding these files will increase the size of the `data` directory from ~2 MB to ~80 MB.  It would be nice to keep both the repository and the PyPI distribution files lightweight if we can, and that big of an increase for data files that many/most installations won't ever touch does seem a bit wasteful.  Keeping the datasets on the Data Hub instead of this repository comes with the downside of having to make network requests during the docs build, which isn't desirable either.

I explored a few alternative formats for `system_50_ac_power_2_full_DST.csv`.  Of course storing as a binary format like parquet helps, as does cutting off unnecessary digits of precision in the power data, and using an encoding for the timestamp index that exploits the consistent 1-minute interval. Results:

- CSV (current form): 50.8 MB
- Zipped CSV: 7.9 MB
- parquet: 14.8 MB
- parquet, float32: 14.2 MB 
- parquet, float32, delta-encoded index: 4.6 MB
- parquet, float32, delta-encoded index, negative values clipped to zero: 3.2 MB

For a dataset with ~1.4 million values (plus as many timestamps), I doubt there's much more opportunity for easy reduction via encoding tweaks.  For reference, here is the code to produce the 3.2 MB file:

```python
df = pd.read_csv("system_50_ac_power_2_full_DST.csv", index_col=0, parse_dates=True)
df.astype(np.float32).clip(lower=0).reset_index().to_parquet("system_50_encoded.parquet",
                                                             use_dictionary=False,
                                                             column_encoding={"measured_on": "DELTA_BINARY_PACKED"})
```

The next question is: do we need the full data file?  It's currently ~2.5 years.  Cutting it down to one year gets the size down to 1.2 MB.  6 months is only 600 kB.  Would shorter data periods still achieve the goals of these examples?  Or resample the file to 5-minute averages?

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+
+# 1) REMOVE STALE DATA (that isn't during nighttime periods)
+# Day/night mask
+daytime_mask = power_or_irradiance(time_series)
+# Stale data mask
+stale_data_mask = gaps.stale_values_round(time_series,
+                                          window=3,
+                                          decimals=2)
+stale_data_mask.loc[(stale_data_mask) &
+                    (~daytime_mask)] = False
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py]: Updated to this in https://github.com/pvlib/pvanalytics/pull/202/commits/54d77682c46faf5dfa56f580055f420c1c93e057

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+# checks (stale or abnormal data). Then we can re-visualize the data
+# post-filtering.
+
+# Filter the time series, taking out all of the issues
+time_series = time_series[~stale_data_mask]
+time_series = time_series[~negative_mask]
+time_series = time_series[erroneous_mask]
+time_series = time_series[~out_of_bounds_mask]
+time_series = time_series[~zscore_outlier_mask]
+time_series = time_series.asfreq(data_freq)
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py]: The asfreq() call at the end of this returns the data to its original frequency, with NaN's for the data that's been filtered out. To make this less sequential, I have updated this to a single call:

`# Filter the time series, taking out all of the issues
issue_mask = ((~stale_data_mask) & (~negative_mask) &
          (~erroneous_mask) & (~zscore_outlier_mask))
time_series = time_series[issue_mask]
time_series = time_series.asfreq(data_freq)`

File: docs/examples/pvfleets-qa-pipeline/pvfleets-temperature-qa.py (hunk 5 above)
Comment (kperrynrel) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-temperature-qa.py]: Did the same here as above to make the mask a single call. asfreq() returns the time series to original data frequency with NaN's:

```python
# Filter the time series, taking out all of the issues
issue_mask = ((~stale_data_mask) & (temperature_limit_mask) &
              (~zscore_outlier_mask))
time_series = time_series[issue_mask]
time_series = time_series.asfreq(data_freq)
```

**Comment by kperrynrel** [comment]
@kandersolar I upsampled all the data so it's 15 minute averaged, and then converted to parquet format. We are now getting an error on parquet support:

 - Missing optional dependency 'pyarrow'. pyarrow is required for parquet support. Use pip or conda to install pyarrow.
 - Missing optional dependency 'fastparquet'. fastparquet is required for parquet support. Use pip or conda to install fastparquet

You ok with me adding either fastpaquet or pyarrow as a dependency? Also, do you have a preference on one?

**Comment by kandersolar** [comment]
> You ok with me adding either fastpaquet or pyarrow as a dependency? Also, do you have a preference on one?

Yep!  I think we can get away with keeping it as a [docs requirement](https://github.com/pvlib/pvanalytics/blob/main/setup.py#L44-L49) rather than an installation requirement.  I'd say mild preference for pyarrow for its maturity.

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (cwhanse) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py]: That's much easier to read, thanks

**Comment by kperrynrel** [comment]
@kandersolar @cwhanse let me know if there's anything else I can update here before the 2.0 release! I updated some of the matplotlib rendering yesterday and am feeling ok with how it is looking. Thanks!

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+    (time_series.index >= data_shift_start_date.tz_convert(
+        time_series.index.tz)) &
+    (time_series.index <= data_shift_end_date.tz_convert(
+        time_series.index.tz))]
+
+time_series = time_series.asfreq(data_freq)
+
+
+# %%
+# Display the final irradiance time series, post-QA filtering.
```
Comment (cwhanse) [inline docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py:361]: This plot isn't rendering

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+time_series.plot(title="Final Filtered Time Series")
+plt.xlabel("Date")
+plt.ylabel("AC Power (kW)")
+plt.tight_layout()
+plt.show()
+plt.close()
```
Comment (cwhanse) [inline docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py:367]:
```suggestion
```

**Review by cwhanse — APPROVED** [review]
Thanks @kperrynrel

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py (hunk 9 above)
Comment (cwhanse) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py:361]: That plot still isn't rendering, idk why not.

**Review by kandersolar — APPROVED** [review]
Thanks @kperrynrel!  One minor comment, otherwise good to merge

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py (hunk 9 above)
Comment (kperrynrel) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py:361]: Looks like it's rendering now, not sure what was going on with it: 
image

File: docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+
+# %%
+# Now, let's run basic data checks to identify stale and abnormal/outlier
+# data in the time series. Basic data checks include the following steps:
+#
+# 1) Flatlined/stale data periods
+#    (:py:func:`pvanalytics.quality.gaps.stale_values_round`)
+# 2) Negative irradiance data
+# 3) "Abnormal" data periods, which are defined as less than 10% of the
+#    daily time series mean OR greater than 1300
```
[the remark this answers is not in the document]
Comment (kperrynrel) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py]: Good catch at @kandersolar, updated docstring to the following in commit https://github.com/pvlib/pvanalytics/pull/202/commits/0cadb6e2d93bf6a2bf05abab4a81e2c0d2cf970b:

# 3) "Abnormal" data periods, which are defined as less than the daily minimum
#    of 50 OR any data greater than 1300

Looks like the docs rendering in the checks for some reason isn't updating, even though the underlying code has been updated. Not sure what's going on with that

[the remark this answers is not in the document]
Comment (kperrynrel) [reply docs/examples/pvfleets-qa-pipeline/pvfleets-irradiance-qa.py]: oops, updated in https://github.com/pvlib/pvanalytics/pull/202/commits/5a01c3aa7b44f752cb3ecf986b32c34b28b1ef27

**Pull request merged**